File: Week_02/cpp/src/_538_ConvertBstToGreaterTree.cpp
#include <iostream>

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};
class UsingDfs
{
public:
    TreeNode *convertBST(TreeNode *root)
    {
        dfs(root, 0);
        return root;
    }

    int dfs(TreeNode *root, int sum)
    {
        if (!root)
            return sum;
        sum = dfs(root->right, sum);
        sum += root->val;
        root->val = sum;
        return dfs(root->left, sum);
    }
};

File: Week_02/kotlin/src/_538_ConvertBstToGreaterTree.kt
import java.util.*

class UsingIterative {
    fun convertBST(root: TreeNode?): TreeNode? {
        var node = root
        val stack = LinkedList<TreeNode>()
        var sum = 0
        while (node != null || stack.isNotEmpty()) {
            while (node != null) {
                stack.push(node)
                node = node.right
            }
            node = stack.pop()
            sum += node.`val`
            node.`val` = sum
            node = node.left
        }
        return root
    }
}